src/Microsoft.ML.Data/Transforms/KeyToVector.cs

```diff
@@ -263,10 +263,11 @@ public Mapper(KeyToVectorMappingTransformer parent, Schema inputSchema)
                 _types = new VectorType[_parent.ColumnPairs.Length];
                 for (int i = 0; i < _parent.ColumnPairs.Length; i++)
                 {
-                    if (_parent._columns[i].Bag || _infos[i].TypeSrc.ValueCount == 1)
-                        _types[i] = new VectorType(NumberType.Float, _infos[i].TypeSrc.ItemType.GetKeyCount());
+                    int valueCount = _infos[i].TypeSrc.GetValueCount();
```

Comment:
>_infos[i].TypeSrc [](start = 37, length = 17)

do you want to cast `_info[i].TypeSrc` to vector first? or code becomes to hard to read?

Reply:
Yes, I think the code is hard to read in that case. It is unfortunate that we have 2 casts here, but I don't think it is a huge deal, do you?

```C#
                for (int i = 0; i < _parent.ColumnPairs.Length; i++)
                {
                    int valueCount;
                    ColumnType srcItemType;
                    if (_infos[i].TypeSrc is VectorType vectorType)
                    {
                        valueCount = vectorType.Size;
                        srcItemType = vectorType.ItemType;
                    }
                    else
                    {
                        valueCount = 1;
                        srcItemType = _infos[i].TypeSrc;
                    }

                    if (_parent._columns[i].Bag || valueCount == 1)
                        _types[i] = new VectorType(NumberType.Float, srcItemType.GetKeyCount());
                    else
                        _types[i] = new VectorType(NumberType.Float, valueCount, srcItemType.GetKeyCount());
                }
```

---
In reply to: [247767549](https://github.com/dotnet/machinelearning/pull/2131#discussion_r247767549) [](ancestors = 247767549)